### PR TITLE
session, tcpip: fix exit-status ordering and half-close in forwarding

### PR DESCRIPTION
**Pull request by kradalby**
This PR brings in the gliderlabs ssh changes from https://github.com/tailscale/tailscale/pull/18331 into our fork.

This is the first part of getting https://github.com/tailscale/tailscale/pull/18331 into main, where the other part will be update the gliderssh dependency and the other half of the changes. 

In addition to the changes, it tries to pick up some of the comments, adds some references to various specifications and tests.

Updates tailscale/tailscale#18256

### CODE REVIEW

File: streamlocal.go [hunk 1]
```diff
@@ -403,11 +403,18 @@ func NewReverseUnixForwardingCallback(opts UnixForwardingOptions) ReverseUnixFor
...
+			listenCtx = ctx
+		}
+
 		cleaned, err := validateSocketPath(socketPath, opts)
 		if err != nil {
 			return nil, err
 		}
-		if opts.PathValidator != nil {
+		// PathValidator requires a valid SSH Context for session
+		// metadata; skip when ctx is nil (only in tests).
```
Comment (neinkeinkaffee) [inline streamlocal.go]: Why do we allow the nil context only for tests instead of just making the test construct one and pass it in?

File: tcpip_test.go [hunk 2]
```diff
@@ -173,3 +174,231 @@ func TestReverseTCPForwardingRespectsCallback(t *testing.T) {
...
+	if !bytes.Equal(buf, msg) {
+		t.Fatalf("got %q, want %q", buf, msg)
+	}
+
+	// Close ext2's write side so bicopy's other direction finishes
+	ext2.(*net.TCPConn).CloseWrite()
+
+	select {
+	case <-done:
+	case <-time.After(5 * time.Second):
```
Comment (neinkeinkaffee) [inline tcpip_test.go]: Nit: 1s should be plenty, do we need 5s?

**Review by neinkeinkaffee — COMMENTED** [review]
Given that we'll cover the exit ordering in an integration test in a follow-up in OSS, this looks good to me apart from one question mark I've got regarding the nil context safeguard!

File: streamlocal.go (hunk 1 above)
Comment (neinkeinkaffee) [reply streamlocal.go]: Are we testing it because we expect that it might happen in real-world usage and we want to allow it? Or is it just convenience for the tests, to reduce setup code?

**Review by neinkeinkaffee — APPROVED** [review]
